_posts/2025-05-09-sorting-for-fast-selective-queries.md

```diff
@@ -0,0 +1,660 @@
+
+</div>
+
+The fastest way to read data is to not read data.
+It's as simple as that!
+This post is all about how to read as little data as possible to answer multiple different patterns of selective read queries.
+
+> Feel free to skip right to the [experiments](#experimental-results)!
+
+The cases where these techniques are most useful are when read performance is more critical than write performance.
+In these cases, we can deliberately add ordering steps that will slow down our writes, but will greatly speed up our reads.
+
+If reading the entire dataset is required for each query in your read workload, then these pruning techniques will not help either!
+However, in practice filters are common, especially to look at recent data points.
+
+Using any kind of sorting will be helpful when:
+
+- Your dataset is large and doesn't fit entirely in memory
+- You access your data via HTTP(S)
+- Your data lives in the cloud on object storage such as AWS S3
+
+These advanced techiques help when one of these situations occur:
+
+- Queries filter on different columns
+- Query patterns are not perfectly predictable
+- Queries filter by time and at least one other column
+
+Cases like those appear often in the wild!
+
+These examples use the DuckDB file format, but thanks to DuckDB's [partial reading support]({% link docs/stable/extensions/httpfs/https.md %}#partial-reading), these techniques can be generally applied to nearly any columnar file format or database.
+This is a great way to speed up querying Parquet files on remote endpoints!
+
+## The “On-Time Flights” Dataset
+
+Ever wondered how likely you will be delayed when you are getting ready to fly?
+If you are traveling in the US, you can use a government dataset to see how on-time each route is!
+You can [get more details on the dataset at its official site](https://www.transtats.bts.gov/DatabaseInfo.asp?QO_VQ=EFD%20&Yv0x=D), but I grabbed a [Parquet version of the data from Kaggle](https://www.kaggle.com/datasets/robikscube/flight-delay-dataset-20182022).
+It includes almost 5 years of data on US flights.
+Combined, the Parquet files are about 1.1 GB.
+This is fairly small when accessed locally but can be more of a challenge over a slow remote connection.
+In the rest of the post, we'll run different benchmarks to demonstrate how pruning can help reducing the amount of network traffic required to answer certain queries.
+
+Our hypothetical use case is to serve a dashboard where people can explore a single `origin`, a single `dest` (destination), or an `origin`–`dest` pair.
+The `origin` and `dest` columns are three-letter airport codes (like `LAX` or `JFK`).
+We assume once you have picked one of those filters, you want all the rows that match and a few arbitrary columns.
+
+Our second experiment will add in a time component and additionally filter to the latest 1, 13, or 52 weeks of data in the dataset.
```

Comment:
This is the first occurrence of the word `experiment`. What is the first experiment?

Reply:
Ah, it was meant to be the first "hypothetical use case" of the dashboard. I will align the terminology!